Automattic/pocket-casts-ios
# Transcripts: search logic

| 📘 Part of: #1848 | 
|:---:|

| iPhone | iPad |
| ------- | ---- |
| ![Simulator Screenshot - iPhone 15 - 2024-08-02 at 15 24 31](https://github.com/user-attachments/assets/7c333e95-fefa-49f6-89fb-87e45064fb97) | ![Simulator Screenshot - iPad Pro (12 9-inch) (6th generation) - 2024-08-02 at 15 24 57](https://github.com/user-attachments/assets/729d309f-3d4c-43e7-92db-76fd14173598) |

This adds the search logic inside a single transcript. I implemented a version of the [Knuth–Morris–Pratt algorithm](https://en.wikipedia.org/wiki/Knuth–Morris–Pratt_algorithm) to be as performant as possible.

## To test

Before testing, ensure **transcripts** feature flag is enabled. It's also better to test on a real device. If you want to test on the simulator, use the simulator keyboard, not your Mac's.

### Testing UI componentes

1. Run the app
2. Play any episode with transcripts, [there's a list here](https://lists.pocketcasts.com/b852a088-ccee-4e4b-a3c5-4bb7fd700a02)
3. Tap the Transcript button and then tap "Search"
4. Scroll down to the end of the transcript
5. ✅ The scroll view should end when the keyboard starts
6. Tap "Done"
7. ✅ The scroll view should get back to its original position
8. Test the same on iPad, but when the keyboard is appearing toggle between portrait/landscape and ensure the scroll view is adjusted accordingly

### Testing search

1. Run the app
2. Play any episode with transcript, [there's a list here](https://lists.pocketcasts.com/b852a088-ccee-4e4b-a3c5-4bb7fd700a02)
3. Tap the transcript button and then tap "Search"
4. Enter any word
5. ✅ The app scrolls to the first results of your search
6. ✅ The input accessory should display "1 of X" where X is the number of results
7. ✅ The first result should have a white highlight color and black text color
8. ✅ The rest of the results should have a white background color with opacity
9. Tap the chevron down or up to toggle between results
10. ✅ The app should scroll to the next result updating which one is highlighted
11. ✅ When you reach the last one and tap the chevron down, it should go to the first
12. ✅ When showing the first result and you tap the chevron up, it should go to the last
13. ✅ The label should update indicating which result you see, eg.: "3 of 10"
14. Tap "search" on the keyboard
15. ✅ The keyboard should disappear, but the results are still highlighter and the input accessory is at the bottom of the screen
16. ✅ Ensure you can still use the chevrons up and down
17. Tap "Done"
18. ✅ All highlights are removed
19. Tap "Search" again and search for something
20. Tap "Done"
21. ✅ Keyboard, input accessory, and highlights should all disappear

## Checklist
